### PR TITLE
KAFKA-14395: add config to configure client supplier

### PR DESCRIPTION
**Pull request by lihaosky**
## Description
This PR is for https://cwiki.apache.org/confluence/display/KAFKA/KIP-884%3A+Add+config+to+configure+KafkaClientSupplier+in+Kafka+Streams. It adds a new `default.client.supplier` config to configure the `KafkaClientSupplier`.

## Test
Unit test

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 1]
```diff
@@ -806,6 +812,11 @@ public class StreamsConfig extends AbstractConfig {
                     LogAndFailExceptionHandler.class.getName(),
                     Importance.MEDIUM,
                     DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_DOC)
+            .define(DEFAULT_CLIENT_SUPPLIER_CONFIG,
+                    Type.CLASS,
+                    DefaultKafkaClientSupplier.class.getName(),
+                    Importance.MEDIUM,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: I think it should be `LOW` priority.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 2]
```diff
@@ -1727,6 +1738,15 @@ public static Set<String> verifyTopologyOptimizationConfigs(final String config)
         return verifiedConfigs;
     }
 
+    /**
+     * Return configured KafkaClientSupplier
+     * @return Configured KafkaClientSupplier
+     */
+    public KafkaClientSupplier getKafkaClientSupplier() {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1770]: Was this part of the KIP and do we need it? `StreamsConfig` is public API and we cannot just change it.

Also, the client supplier does not implement `Configurable` and thus does not need to be configured.

I am ok with reusing `getConfiguredInstance` but we would need a small change to the KIP.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 3]
```diff
@@ -1088,6 +1088,28 @@ public void shouldTriggerRecordingOfRocksDBMetricsIfRecordingLevelIsDebug() {
...
 
+    @Test
+    public void shouldGetClientSupplierFromConfigForConstructor() {
+        final StreamsConfig config = new StreamsConfig(props);
+        final StreamsConfig mockConfig = spy(config);
+        when(mockConfig.getKafkaClientSupplier()).thenReturn(supplier);
+
+        new KafkaStreams(getBuilderWithSource().build(), mockConfig);
+        // It's called once in above when mock
+        verify(mockConfig, times(2)).getKafkaClientSupplier();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:1099]: Why `times(2)` but not `times(1)` ?

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 4]
```diff
@@ -1088,6 +1088,28 @@ public void shouldTriggerRecordingOfRocksDBMetricsIfRecordingLevelIsDebug() {
         }
     }
 
+    @Test
+    public void shouldGetClientSupplierFromConfigForConstructor() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:1092]: Should we also test, that a supplier passed into the constructor overwrites the config?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 2 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1770]: Sure. Will update the KIP.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 4 above)
Comment (lihaosky) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:1092]: Sure.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 3 above)
Comment (lihaosky) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:1099]: Somehow above `when(mockConfig.getKafkaClientSupplier()).thenReturn(supplier);` is counted.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 5]
```diff
@@ -1076,7 +1082,12 @@ public class StreamsConfig extends AbstractConfig {
                     Type.LONG,
                     null,
                     Importance.LOW,
-                    WINDOW_SIZE_MS_DOC);
+                    WINDOW_SIZE_MS_DOC)
+            .define(DEFAULT_CLIENT_SUPPLIER_CONFIG,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: nit: we keep the configs grouped by priority and order alphabetically within each group. Can we more this accordingly?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1770]: Thanks. Can you send a follow up email to the vote thread, calling out the change?

We don't need to re-vote but if anybody would have an objection, we need to give them a chance to race it.

**Review by mjsax — COMMENTED** [review]
Overall LGTM.

Can merge this after you send the follow up email and let a few days pass (and if nobody raises concerns)

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 5 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Sure

**Comment by mjsax** [comment]
Merged to `trunk`. Thanks for the PR and KIP.

**Pull request merged**